Framework/Geometry/test/MeshObjectTest.h

```diff
@@ -0,0 +1,1089 @@
+        boost::shared_ptr<MeshObject>(new MeshObject);
+    retVal->initialize(triangles, vertices);
+    return retVal;
+  }
+
+  boost::shared_ptr<MeshObject> createLShape() {
+    /**
+    * Create an L shape with vertices at
+    * (0,0,Z) (2,0,Z) (2,1,Z) (1,1,Z) (1,2,Z) & (0,2,Z),
+    *  where Z = 0 or 1.
+    */
+    std::vector<V3D> vertices;
+    vertices.push_back(V3D(0, 0, 0));
+    vertices.push_back(V3D(2, 0, 0));
+    vertices.push_back(V3D(2, 1, 0));
+    vertices.push_back(V3D(1, 1, 0));
+    vertices.push_back(V3D(1, 2, 0));
+    vertices.push_back(V3D(0, 2, 0));
+    vertices.push_back(V3D(0, 0, 1));
+    vertices.push_back(V3D(2, 0, 1));
+    vertices.push_back(V3D(2, 1, 1));
+    vertices.push_back(V3D(1, 1, 1));
+    vertices.push_back(V3D(1, 2, 1));
+    vertices.push_back(V3D(0, 2, 1));
+
+    std::vector<int> triangles;
+    // z min
+    triangles.push_back(0);
+    triangles.push_back(5);
+    triangles.push_back(1);
+    triangles.push_back(1);
+    triangles.push_back(3);
+    triangles.push_back(2);
+    triangles.push_back(3);
+    triangles.push_back(5);
+    triangles.push_back(4);
+    // z max
+    triangles.push_back(6);
+    triangles.push_back(7);
+    triangles.push_back(11);
+    triangles.push_back(11);
+    triangles.push_back(9);
+    triangles.push_back(10);
+    triangles.push_back(9);
+    triangles.push_back(7);
+    triangles.push_back(8);
+    // y min
+    triangles.push_back(0);
+    triangles.push_back(1);
+    triangles.push_back(6);
+    triangles.push_back(6);
+    triangles.push_back(1);
+    triangles.push_back(7);
+    // x max
+    triangles.push_back(1);
+    triangles.push_back(2);
+    triangles.push_back(7);
+    triangles.push_back(7);
+    triangles.push_back(2);
+    triangles.push_back(8);
+    // y mid
+    triangles.push_back(2);
+    triangles.push_back(3);
+    triangles.push_back(8);
+    triangles.push_back(8);
+    triangles.push_back(3);
+    triangles.push_back(9);
+    // x mid
+    triangles.push_back(3);
+    triangles.push_back(4);
+    triangles.push_back(9);
+    triangles.push_back(9);
+    triangles.push_back(4);
+    triangles.push_back(10);
+    // y max
+    triangles.push_back(4);
+    triangles.push_back(5);
+    triangles.push_back(10);
+    triangles.push_back(10);
+    triangles.push_back(5);
+    triangles.push_back(11);
+    // x min
+    triangles.push_back(5);
+    triangles.push_back(0);
+    triangles.push_back(11);
+    triangles.push_back(11);
+    triangles.push_back(0);
+    triangles.push_back(6);
+
+    boost::shared_ptr<MeshObject> retVal =
+        boost::shared_ptr<MeshObject>(new MeshObject);
+    retVal->initialize(triangles, vertices);
+    return retVal;
+  }
+};
+
+// -----------------------------------------------------------------------------
+// Performance tests
+// -----------------------------------------------------------------------------
+class MeshObjectTestPerformance : public CxxTest::TestSuite {
```

Comment:
👍 for thinking to include performance tests. Could you please.

1. Ensure that all the performance critical behaviour is covered. such as `isValid`, `isOnSide`, see public methods on `IObject` a guide.
1. Ensure that the test run durations are reasonable in length (~1/10th to 1 second) per test method so that we have a good ability to track performance changes.

Reply:
The current performance test is similar to the test used by CSGObject (was object).

Reply:
I regard the extension of the performance tests beyond that of CSGObject beyond the scope of this issue. The current test covers isValid and intersection.

Reply:
We need a baseline to start tracking performance from and this means putting the tests in place so that we can begin tracking. Lamar Moore tracks the performance across the framework weekly looking for regressions.  As the implementation of math for `MeshObject` has been provided completely bespoke, I consider the performance tests to be very important.  I don't consider `CSGObjectTestPerformance` as a good example to follow when it comes to complete performance tests.

Reply:
I think a MeshObject with a large number vertices and triangles would be needed for this and am think of an algorithm to generate one.

Reply:
A better approach would be to call the method lots of times to build up enough statistics.

Reply:
I was thinking that there might be performance issues that only arise on a larger mesh.

Reply:
I don't think that's such a concern. The way your are storing the information looks quite efficient to me. However you have had to implement all aspects of the mesh related math yourself, and that's worth covering with performance tests.